---
title: "CNA dashboard"
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: fill
    theme: default
    logo: navin_lab_logo.png
params:
  directory:
    value:
  outdir: "output"
  facs:
editor_options: 
  chunk_output_type: console
---

``` {js}
// Inverse color of navigation bar.
$('.navbar-inverse').removeClass('navbar-inverse').addClass('navbar-default');
```


```{r setup, include=FALSE, message=FALSE}
# checking for packages, loading or installing
packages <- c("tidyverse", "fs", "shiny",
              "here", "flexdashboard", "cowplot",
              "janitor", "ape", "ggsci", "plotly",
              "amap", "paletteer", "scales", "umap","uwot",
              "flsa", "BiocManager", "DT", "devtools",
              "flowCore","splus2R","locfit","dplyr","plotly")

package.check <- lapply(packages, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    install.packages(x, dependencies = TRUE,repos = "http://cran.us.r-project.org")
    suppressWarnings(library(x, character.only = TRUE))
  }
})

bioc_packages <- c("ggtree", "ComplexHeatmap")

lapply(bioc_packages, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    BiocManager::install(x, update = FALSE)
    suppressWarnings(library(x, character.only = TRUE))
  }
})

git_packages<- c("Rphenograph")
lapply(git_packages, FUN = function(x) {
if(!require(x, character.only = TRUE)){
  devtools::install_github("JinmiaoChenLab/Rphenograph")
  library(x,character.only = TRUE)
}

})

theme_set(theme_cowplot())
```

```{r setting_up, include=FALSE}
# message
message("Rendering Dashboard")

# reading data
dat_seg <- readr::read_tsv(fs::dir_ls(path = paste0(here(),"/",params$outdir), recurse = T, glob= "*uber*seg.txt")) %>% janitor::clean_names()
dat_seg_cp <- dat_seg %>% dplyr::select(-chrom, -chrompos, -abspos)

dat_bin <- readr::read_tsv(fs::dir_ls(path = paste0(here(),"/",params$outdir), recurse = T, glob= "*uber*bin.txt")) %>% janitor::clean_names()
dat_stats <- readr::read_tsv(fs::dir_ls(path = paste0(here(),"/",params$outdir), recurse = T, regex= "all_stat_metrics.txt")) %>% 
  janitor::clean_names() %>% 
  mutate(duplicates = dups_removed/total_reads)
Raw_dat_stats<-readr::read_tsv(fs::dir_ls(path = paste0(here(),"/",params$outdir), recurse = T, regex= "all_stat_metrics.txt"))

# correcting names
dat_stats$sample_name <- make_clean_names(dat_stats$sample_name)

dat_stats <- dat_stats %>%
  dplyr::filter(sample_name %in% names(dat_seg))

# dat_stats metrics
dat_stats_metrics <- dat_stats %>% 
  dplyr::select(-sample_name) %>% 
  summarise_all(median)

# reading ratio
dat_rat <- readr::read_tsv(fs::dir_ls(path = paste0(here(),"/",params$outdir), recurse = T, glob= "*uber*ratio.txt")) %>% 
  janitor::clean_names()
dat_rat_cp <- dat_rat %>% dplyr::select(-chrom, -chrompos, -abspos)

# dat_seg setup
dat_seg_s <- dat_seg %>% 
  dplyr::select(-chrom, -chrompos, - abspos)

dat_seg_t <- as.data.frame(t(dat_seg_s))

# n cells
n_cells <- dat_seg %>% 
  dplyr::select(-abspos, 
                -chrompos, 
                -chrom) %>% 
  ncol()

# sourcing integer inference
source(paste(params$directory, "/integer_cn.R", sep = ""))
source(paste(params$directory, "/fac_peak.R", sep = ""))
```

```{r clustering, include = FALSE}
if (n_cells < 30) {
  umap_n_neighbors <- n_cells - 1
} else umap_n_neighbors <- 30

set.seed(42)
dat_umap <- uwot::umap(dat_seg_t, 
                       metric = "manhattan",
                       n_threads = 20,
                       min_dist = 0,
                       n_neighbors = umap_n_neighbors)
  
umap_df <- as.data.frame(dat_umap) %>% 
  dplyr::rename("umap1" = "V1",
                "umap2" = "V2")
rownames(umap_df) <- rownames(dat_seg_t)

if (n_cells < 30) {
  k_param <- n_cells - 1
} else k_param <- 30
  
pheno <- Rphenograph::Rphenograph(umap_df, k = k_param)
cl <- tibble(cell = rownames(umap_df),
             cluster = igraph::membership(pheno[[2]])) %>% 
  arrange(cluster)

umap_df <- umap_df[cl$cell,]

# getting number of needed colors for cl
n_colors <- cl$cluster %>% unique() %>% length()

# ordering cells
dat_seg_order <- dat_seg_t[cl$cell,]

# setting colors
row_col <- setNames(colorRampPalette(paletteer_d(pals, 
                                alphabet, 26))(n_colors),
                    levels(as.factor(cl$cluster)))
```

Summary
=======================================================================

Column {data-width=200}
-----------------------------------------------------------------------

### Cells Sequenced {.colored}

```{r}
 Raw_dat_stats%>% 
  nrow() %>% 
  valueBox(icon = "circle")
```


### Cells Analysed{.colored}

```{r}
dat_seg %>% 
  dplyr::select(-chrom, -chrompos, -abspos) %>% 
  ncol() %>% 
  valueBox(icon = "circle")
```

### Total Reads (Median) {.colored}

```{r tot_reads_val_box}
dat_stats_metrics$total_reads %>% 
  valueBox(icon = "percentage")
```

### Kept Reads (Median) {.colored}

```{r kep_reads_val_box}
dat_stats_metrics$reads_kept %>% 
  valueBox(icon = "percentage")
```

### Bin Count (Median) {.colored}

```{r bincount_val_box}
dat_stats_metrics$median_bin_count %>% 
  valueBox(icon = "percentage")
```

### PCR Duplicates (Median) {.colored}

```{r calculating_pcr_duplicates}
dups <- round(dat_stats_metrics$duplicates*100,2)
valueBox(paste0(dups,"%"))
```

### Clusters
```{r n_clusters}
cl$cluster %>% unique() %>% length() %>% 
  valueBox()
```

Column
-----------------------------------------------------------------------

### Heatmap 

```{r making_heatmap}
# chromosome annotation top bar
# getting the vector of chrom lengths
chr_lengths <- dat_seg %>% 
  dplyr::select(abspos, chrom, chrompos) %>% 
  group_by(chrom) %>%
  summarize(n = n()) %>%
  pull(n)

chr_binary <- rep(c(2,1), 12)
chr <- data.frame(chr = rep.int(x = chr_binary, times = chr_lengths))
# getting lengths for chr numbers annotation
chr_rl_c <- c(1, cumsum(chr_lengths))
# creating a data frame to calculate rowMeans
chr_df <-  data.frame(a = chr_rl_c[1:length(chr_rl_c)-1],b= chr_rl_c[2:length(chr_rl_c)])
chr_l_means <- round(rowMeans(chr_df))
chrom.names <- c(1:22,"X", "Y")
# creating the vector for chr number annotations
v <- vector(mode = "character",length = cumsum(chr_lengths)[length(chr_lengths)])
v[chr_l_means] <- chrom.names
v[is.na(v)] <- ""
chr_bar <- HeatmapAnnotation(chr_text = anno_text(v, 
                                                  gp = gpar(fontsize = 8)),
                             df = chr,
                             show_legend = FALSE,
                             which = "column",
                             col = list(chr = c("1" = "grey88", "2" = "black"))
)

heat_row_col <- rowAnnotation(df =
  data.frame(clusters = sort(cl$cluster)),
             col = list(clusters = row_col))

ht <- Heatmap(as.matrix(log2(dat_seg_order+1e-3)),
              cluster_columns = FALSE,
              cluster_rows = FALSE,
              show_row_names = FALSE,
              show_column_names = FALSE,
              use_raster = TRUE,
              top_annotation = chr_bar,
              heatmap_legend_param = list(title = "Log2 (Ratio)"))

draw(heat_row_col + ht)
```


Column 
-----------------------------------------------------------------------

### UMAP. Hover mouse for more info. {data-height=500}

```{r generating_umap}
p_umap <- umap_df %>% 
  rownames_to_column(var = "Cell") %>%
  dplyr::mutate(cluster = cl$cluster) %>% 
  ggplot() +
  geom_point(aes(x = umap1, 
                 y = umap2, 
                 color = as.factor(cluster),
                 label = Cell),
             show.legend = FALSE) +
  scale_color_manual(values = row_col) +
  theme(legend.position = "none",
        panel.grid = element_blank(),
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank()) 

ggplotly(p_umap, tooltip = "Cell")
```


### FACs Ploidy Inference. Hover mouse for more info. {data-height=500}

```{r infering_ploidies}
table_facs<-facs.density(params$facs)
names(table_facs)<-c("DAPI fluorescence","Density","is_peak","peak_rank")
#ggplot(table_facs)+geom_line(aes(x=`DAPI fluorescence`, y=density ))
facs_peaks <- table_facs[table_facs$is_peak,]
infered_ploidy<-2 * facs_peaks[facs_peaks$peak_rank==2,]$`DAPI fluorescence` / facs_peaks[facs_peaks$peak_rank==1,]$`DAPI fluorescence`
graph <- ggplot(table_facs, aes(x = `DAPI fluorescence`, y = Density)) +
  geom_line(color='steelblue') +
  ggtitle(paste0("FACS Ploidy=",round(infered_ploidy,2))) +
  theme_classic() +
  theme(
    legend.position = 'none'
  )

library("plotly")
ggplotly(graph)
```

QC
=======================================================================

Row {.tabset .tabset-fade data-width=400}
-----------------------------------------------------------------------

### Reads Kept

```{r}
# creates a index for x and y so I have a square matrix, also creates a lable for plotly
p <- dat_stats %>% 
  mutate(idx_x = rep(1:sqrt(nrow(dat_stats)), length.out = nrow(dat_stats), each = sqrt(nrow(dat_stats)))) %>%
  mutate(idx_y = rep(paste(1:sqrt(nrow(dat_stats)),"a", sep = ""), length.out = nrow(dat_stats)))  %>% 
  dplyr::mutate(Cell = paste(sample_name,
                              "- Reads Kept: ",
                             reads_kept,
                             sep = "")) %>% 
  ggplot() + 
  geom_tile(aes(x = idx_x, 
                y = idx_y, 
                fill = reads_kept,
                label = Cell)) +
  coord_equal() + 
  scale_fill_viridis_c(option = "magma",
                       labels = scales::comma) +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank()) +
  ylab("") +
  xlab("") 


# ggplotly
p
# ggplotly(p, tooltip = "Cell")
```

### Duplicates percentage

```{r}
# creates a index for x and y so I have a square matrix, also creates a lable for plotly
p_dups <- dat_stats %>% 
  mutate(idx_x = rep(1:sqrt(nrow(dat_stats)), 
                     length.out = nrow(dat_stats), 
                     each = sqrt(nrow(dat_stats)))) %>%
  mutate(idx_y = rep(paste(1:sqrt(nrow(dat_stats)),"a", sep = ""),
                     length.out = nrow(dat_stats)))  %>% 
  dplyr::mutate(Cell = paste(sample_name,
                              "- Duplicates: ",
                             scales::percent(duplicates),
                             sep = "")) %>% 
  ggplot() + 
  geom_tile(aes(x = idx_x, 
                y = idx_y, 
                fill = duplicates,
                label = Cell)) +
  coord_equal() + 
  scale_fill_viridis_c(option = "viridis",
                       labels = scales::percent) +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank()) +
  ylab("") +
  xlab("") 


# ggplotly
p_dups
# ggplotly(p_dups, tooltip = "Cell") 
```

### Bin read count hist

```{r}
dat_bin %>% 
  dplyr::select(-chrom, -chrompos, -abspos) %>% 
  purrr::map_dfr(median) %>% 
  gather(key = "cell",
         value = "bin_median") %>% 
  dplyr::filter(bin_median > 1) %>% 
  ggplot() +
  geom_histogram(aes(x = bin_median, 
                     y= ..count../sum(..count..)),
                 fill = "white",
                 color = "black") +
  xlab("GC normalized median bin count") +
  ylab("Frequency")
```

### Reads Hist

```{r}
dat_stats %>% 
  gather(key = "parameter",
         value = "value",
         -sample_name) %>% 
  dplyr::filter(parameter %in% c("total_reads", "reads_kept"),
                value > 500) %>% 
  ggplot() +
  geom_histogram(aes(x = value,
                     y = ..count../sum(..count..),
                     fill = parameter),
                 color = "black") +
  facet_wrap(vars(parameter)) +
  scale_fill_manual(values = c("#7D3232", "#32324B")) +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 7)) +
  theme(axis.text.x = element_text(angle = 90),
        strip.background = element_rect(fill= NA),
        legend.position = "none") +
  xlab("Reads") +
  ylab("Frequency")
```

Row
-----------------------------------------------------------------------

### Reads information

```{r}
dt_dat_stats <- dat_stats %>% 
  arrange(desc(reads_kept)) %>% 
  mutate(duplicates = scales::percent(round(duplicates,2)))

DT::datatable(dt_dat_stats, options = list(pageLength = 50))
```


Phylogenetics
=======================================================================

Row
-----------------------------------------------------------------------

### Neighbor-Joining tree

```{r running_phylogenetics}
tree <- ape::nj(
  amap::Dist(dat_seg_order, 
             method = "manhattan", 
             nbproc = 40)
)

# obtaining colors
list_samples <- split(rownames(dat_seg_order),
                      sort(cl$cluster))
tree <- ggtree::groupOTU(tree, list_samples) 

ggtree(ladderize(tree),
  ladderize = FALSE,
  size = .1) +
  geom_tippoint(aes(color = group), size = .7) +
  scale_colour_manual(values = c("black", row_col)) +
  theme(legend.position = "right") +
  labs(color = "Cluster")
```

About
=======================================================================

Row
-----------------------------------------------------------------------

### Dashboard
This dashboard was developed by Darlan Conterno Minussi and is a part of the Navin Lab copy number pipeline.

### Clustering
Clustering of the data is being performed with UMAP dimension reduction followed by R phenograph.

UMAP is using the following parameters:

- Seed: 42
- Distance: Manhattan
- min_dist: 0
- n_neighbors: 30 (if n_cells < 30 then k = n_cells - 1)

RPhenograph is using the following parameters:

- k: 30 (if n_cells < 30 then k = n_cells - 1)

### Ploidy Inference
Ploidy inference is using the fourier transform method developed by Alexander Davis. For more information please refer to the following publication.
Ref: Soon

### Phylogenetics
Phylogenetics is using the Neighbor-Joining method and the distance matrix is calculated using manhattan distance.

Row
-----------------------------------------------------------------------

### Packages
The following packages were used in this dashboard:

- flexdashboard
- tidyverse
- here
- fs
- cowplot
- DT
- shiny
- janitor
- ape
- ggtree
- ggsci
- plotly
- amap
- paletteer
- scales
- uwot
- dbscan
- ComplexHeatmap
- flsa
- Rphenograph

