---
title: "CNA dashboard"
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: fill
    theme: default
    logo: navin_lab_logo.png
params:
  outdir:
  id:
editor_options: 
  chunk_output_type: console
---

``` {js}
// Inverse color of navigation bar.
$('.navbar-inverse').removeClass('navbar-inverse').addClass('navbar-default');
```


```{r setup, include=FALSE, message=FALSE}
# checking for packages, loading or installing
packages <- c("tidyverse", "fs", "shiny",
              "here", "flexdashboard", "cowplot",
              "janitor", "ape", "ggsci", "plotly",
              "amap", "paletteer", "scales", "umap","uwot",
              "flsa", "BiocManager", "DT", "devtools", "Metrics","scquantum")
package.check <- lapply(packages, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    install.packages(x, dependencies = TRUE,repos = "http://cran.us.r-project.org")
    suppressWarnings(library(x, character.only = TRUE))
  }
})

bioc_packages <- c("ggtree", "ComplexHeatmap")

lapply(bioc_packages, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    BiocManager::install(x, update = FALSE)
    suppressWarnings(library(x, character.only = TRUE))
  }
})

git_packages<- c("Rphenograph")
lapply(git_packages, FUN = function(x) {
if(!require(x, character.only = TRUE)){
  devtools::install_github("JinmiaoChenLab/Rphenograph")
  library(x,character.only = TRUE)
}

})

options(bitmapType='cairo')
theme_set(theme_cowplot())
```

```{r setting_up,include=FALSE}
# message
message("Rendering Dashboard")

# reading data
path=paste0(dirname(normalizePath(params$outdir)),"/")
message(path)
message(getwd())
dat_bin <- read.table(system(paste0("find ",here(),"/",params$outdir," -name \"*.bin.txt\" -type f"), intern = TRUE),sep="\t",header=T) %>% janitor::clean_names()
message(dim(dat_bin))
dat_seg <- read.table(system(paste0("find ",here(),"/",params$outdir," -name \"*.seg.txt\" -type f"), intern = TRUE),sep="\t",header=T) %>% janitor::clean_names()
dat_seg_cp <- dat_seg %>% dplyr::select(-chrom, -chrompos, -abspos)
message(dim(dat_seg))

dat_stats <- read.table(system(paste0("find ",here(),"/",params$outdir," -name \"*stat\\_metrics.txt\" -type f"), intern = TRUE),sep="\t",header=T) %>%
  janitor::clean_names() %>%
  mutate(duplicates = dups_removed/total_reads)
message(dim(dat_stats))

# correcting names
dat_stats$sample_name <- make_clean_names(dat_stats$sample_name)

raw_dat_stats <- dat_stats

dat_stats <- dat_stats %>%dplyr::filter(sample_name %in% names(dat_seg))

dat_stats_rm <- raw_dat_stats %>% 
  dplyr::filter(!sample_name %in% names(dat_seg))

# dat_stats metrics
dat_stats_metrics <- dat_stats %>% dplyr::select(-sample_name) %>%summarise_all(median)
# reading ratio

dat_rat <- read.table(system(paste0("find ",here(),"/",params$outdir," -name \"*.ratio.txt\" -type f"), intern = TRUE),sep="\t",header=T) %>%janitor::clean_names()
#dat_rat_cp <- dat_rat %>% dplyr::select(-chrom, -chrompos, -abspos)

message(dim(dat_rat))
# dat_seg setup
dat_seg_s <- dat_seg %>%dplyr::select(-chrom, -chrompos, -abspos)

dat_seg_t <- as.data.frame(t(dat_seg_s))
#dat_seg_t[1:5,1:5]

# n cells
n_cells <- dat_seg %>%dplyr::select(-abspos, -chrompos,-chrom) %>%ncol()

```


```{r clustering, include = FALSE}
if (n_cells < 30) {
  umap_n_neighbors <- n_cells - 1
} else umap_n_neighbors <- 30

set.seed(42)
dat_umap <- uwot::umap(dat_seg_t, 
                       metric = "manhattan",
                       n_threads = 20,
                       min_dist = 0,
                       n_neighbors = umap_n_neighbors)
  
umap_df <- as.data.frame(dat_umap) %>% 
  dplyr::rename("umap1" = "V1",
                "umap2" = "V2")
rownames(umap_df) <- rownames(dat_seg_t)

if (n_cells < 30) {
  k_param <- n_cells - 1
} else k_param <- 30
  
pheno <- Rphenograph::Rphenograph(umap_df, k = k_param)
cl <- tibble(cell = rownames(umap_df),
             cluster = igraph::membership(pheno[[2]])) %>% 
  arrange(cluster)

umap_df <- umap_df[cl$cell,]

# getting number of needed colors for cl
n_colors <- cl$cluster %>% unique() %>% length()
# ordering cells
dat_seg_order <- dat_seg_t[cl$cell,]

# setting colors
row_col <- setNames(colorRampPalette(paletteer_d("pals::alphabet", 26))(n_colors),
                    levels(as.factor(cl$cluster)))
```

Summary
=======================================================================

Column {data-width=200}
-----------------------------------------------------------------------

### Sample ID

```{r}
 params$id %>%
  valueBox(icon = "circle",
           color = "#000000")
```


### Cells Sequenced {.colored}

```{r}
 raw_dat_stats %>% 
  nrow() %>% 
  valueBox(icon = "circle",
           color = "#000000")
```

### Cells removed
```{r}
total <- raw_dat_stats %>% nrow() 
kept_number <- dat_seg %>% dplyr::select(-chrom, -chrompos, -abspos) %>% 
  ncol() 

rm = total - kept_number

valueBox(rm, 
         color = "#ff0000")
```

### Cells passed QC {.colored}

```{r}
dat_seg %>% 
  dplyr::select(-chrom, -chrompos, -abspos) %>% 
  ncol() %>% 
  valueBox(icon = "circle",
           color = "#008000")
```

### Total Reads (All PF Reads) {.colored}

```{r tot_reads_val_box_all}
sum(dat_stats$total_reads) %>% 
  valueBox(icon = "percentage")
```


### Total Reads Per Cell(Median) {.colored}

```{r tot_reads_val_box}
dat_stats_metrics$total_reads %>% 
  valueBox(icon = "percentage")
```

### Kept Reads Per Cell (Median) {.colored}

```{r kep_reads_val_box}
dat_stats_metrics$reads_kept %>% 
  valueBox(icon = "percentage")
```

### Bin Count (Median) {.colored}

```{r bincount_val_box}
dat_stats_metrics$median_bin_count %>% 
  valueBox(icon = "percentage")
```

### PCR Duplicates (Median) {.colored}

```{r calculating_pcr_duplicates}
dups <- round(dat_stats_metrics$duplicates*100,2)
valueBox(paste0(dups,"%"))

```


### Median Ploidy (scquantum) 
```{r infering_ploidies}

all.values <- apply(dat_bin[,-(1:3)], 2, ploidy.inference, chrom=dat_bin$chrom)
df<-data.frame(sample_name=names(all.values),
           ploidy=sapply(all.values, function(x) x$ploidy),
           peak.height=sapply(all.values, function(x) x$peak_height)) %>%
            dplyr::mutate(Cell = paste(sample_name,
                              ": ",
                             ploidy,
                              sep = ""))
df$ploidy[is.na(df$ploidy)]<-0
ploidy_median <- median(df$ploidy)
round(ploidy_median,2) %>% valueBox(icon = "percentage")

```


### Clusters
```{r n_clusters}
cl$cluster %>% unique() %>% length() %>% 
  valueBox()
```


Column
-----------------------------------------------------------------------

### Heatmap 


```{r making_heatmap, fig.height=9}


# chromosome annotation top bar
# getting the vector of chrom lengths
chr_lengths <- dat_seg %>% 
  dplyr::select(abspos, chrom, chrompos) %>% 
  group_by(chrom) %>%
  dplyr::summarise(n = n()) %>%
  dplyr::pull(n)

chr_binary <- rep(c(2,1), 12)
chr <- data.frame(chr = rep.int(x = chr_binary, times = chr_lengths))
# getting lengths for chr numbers annotation
chr_rl_c <- c(1, cumsum(chr_lengths))
# creating a data frame to calculate rowMeans
chr_df <-  data.frame(a = chr_rl_c[1:length(chr_rl_c)-1],b= chr_rl_c[2:length(chr_rl_c)])
chr_l_means <- round(rowMeans(chr_df))
chrom.names <- c(1:22,"X", "Y")
# creating the vector for chr number annotations
v <- vector(mode = "character",length = cumsum(chr_lengths)[length(chr_lengths)])
v[chr_l_means] <- chrom.names
v[is.na(v)] <- ""
chr_bar <- HeatmapAnnotation(chr_text = anno_text(v, 
                                                  gp = gpar(fontsize = 8)),
                             df = chr,
                             show_legend = FALSE,
                             which = "column",
                             col = list(chr = c("1" = "grey88", "2" = "black"))
)

Peak=rep("Peak_D",nrow(dat_seg_order))
Peak[grep(rownames(dat_seg_order),pattern = "_A_",ignore.case = T)]="Peak_A"
Peak[grep(rownames(dat_seg_order),pattern = "A_",ignore.case = T)]="Peak_A"
Peak[grep(rownames(dat_seg_order),pattern = "Ctrl",ignore.case = T)]="Ctrl"

Cells=rep("preTX",nrow(dat_seg_order))
Cells[grep(rownames(dat_seg_order),pattern = "preTX",ignore.case = T)]="MIDTX"
col<-c("blue","orange")
names(col)<-c("MIDTX", "preTX")
pcol<-c("red","purple","green")
names(pcol)<-c("Ctrl","Peak_D","Peak_A")

heat_row_col <- rowAnnotation(
	     df = as.data.frame(Peak),
             col = list(Peak = pcol))
ht <- Heatmap(as.matrix(log2(dat_seg_order+1e-3)),
              cluster_columns = FALSE,
              cluster_rows = TRUE,
              show_row_names = FALSE,
              show_column_names = FALSE,
              use_raster = TRUE,
	      left_annotation = heat_row_col,
              top_annotation = chr_bar,
	            col = circlize::colorRamp2(breaks = c(-2,0.1,2),c("dodgerblue3", "white", "firebrick3")),
              heatmap_legend_param = list(title = "Log2 (Ratio)", title_gp = gpar(fontsize = 15, fontface = "bold"),labels_gp = gpar(fontsize = 12)))


draw(ht,heatmap_legend_side = "right")
```


Column 
-----------------------------------------------------------------------

### UMAP. Hover mouse for more info. {data-height=500}

```{r generating_umap}
p_umap <- umap_df %>% 
  rownames_to_column(var = "Cell") %>%
  dplyr::mutate(cluster = cl$cluster) %>% 
  ggplot() +
  geom_point(aes(x = umap1, 
                 y = umap2, 
                 color = as.factor(cluster),
                 label = Cell),
             show.legend = FALSE) +
  #scale_color_manual(values = row_col) +
  theme(legend.position = "none",
        panel.grid = element_blank(),
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank()) 

ggplotly(p_umap, tooltip = "Cell")
```


### Ploidy Inference. Hover mouse for more info. {data-height=500}

```{r draw_ploidies}



# obtaining clusters labels to use as colors, matching with cell name and plotting
umap_cl_result <- umap_df %>% rownames_to_column("sample_name") %>%  mutate(cluster = cl$cluster) %>% dplyr::select(sample_name, cluster) 

dat_ploidy_cl <- dplyr::inner_join(df,umap_cl_result, by = "sample_name")

p_ploidy <- ggplot(dat_ploidy_cl,aes(x = ploidy, 
                 y = peak.height, label = Cell)) + 
  geom_point(color = as.factor(dat_ploidy_cl$cluster)) +
 # scale_color_manual(values = row_col) + 
  scale_x_continuous(breaks = scales::pretty_breaks()) +
  scale_y_continuous(breaks = scales::pretty_breaks()) +
  xlab("Inferred ploidy") +
  ylab("Peak Height (Confidence)") + 
  labs(color = "Cluster")

ggplotly(p_ploidy, tooltip = "Cell")
```


QC
=======================================================================

Row {.tabset .tabset-fade data-width=400}
-----------------------------------------------------------------------

### Quality
```{r}
message("Calculating RMSE")

rmses <- vector()

for (i in 4:ncol(dat_seg)) {
  actual <- as.tibble(dat_seg[, i]) %>% dplyr::pull()
  predicted <- as.tibble(dat_rat[, i]) %>% dplyr::pull()

  rmses[i] <- rmse(actual, predicted)
  names(rmses)[i] <- names(dat_seg)[i]
}

rmses_df <- enframe(rmses,
  name = "sample_name",
  value = "rmse"
)

# cv
message("Calculating CV")
cv <- vector()

for(i in 4:ncol(dat_seg)) {
  seg_n_probes <- rle(pull(as.tibble(dat_seg[,i])))$lengths
  ratio_split <- split(pull(as.tibble(dat_rat[,i])), 
                       as.factor(rep.int(1:length(seg_n_probes), seg_n_probes)))
  cv[i] <- lapply(ratio_split, function(x) sd(x)/mean(x)) %>% unlist() %>% sum()
  names(cv)[i] <- names(dat_seg)[i]
}

cv_df <- enframe(cv,
                      name = "sample_name",
                      value = "cv"
)

cv_rmse_df <- dplyr::inner_join(rmses_df, cv_df)

p <- cv_rmse_df %>% 
  ggplot(aes(cv, rmse)) +
  geom_point(aes(label = sample_name)) +
  geom_smooth(method = "lm") +
  #viridis::scale_color_viridis(discrete = TRUE) +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)) +
  scale_y_continuous(breaks = scales::pretty_breaks(n = 10)) +
  xlab("Coefficient of Variation") + 
  ylab("RMSE")

p <- ggplotly(p)

p_rmse <- dplyr::inner_join(dat_stats, cv_rmse_df) %>% 
  ggplot(aes(reads_kept, rmse)) + 
  geom_point() +
  geom_smooth() +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)) + 
  scale_y_continuous(breaks = scales::pretty_breaks(n = 10)) + 
 # scale_color_viridis_c() +
  theme(axis.text.x = element_text(angle = 90)) +
  ylab("RMSE")
  

p_rmse <- ggplotly(p_rmse)

subplot(p,p_rmse,
        nrows = 2,
        titleX = TRUE,
        titleY = TRUE)

```



### Bin count

```{r}
dat_bin %>% 
  dplyr::select(-chrom, -chrompos, -abspos) %>% 
  purrr::map_dfr(median) %>% 
  gather(key = "cell",
         value = "bin_median") %>% 
  dplyr::filter(bin_median > 1) %>% 
  ggplot() +
  geom_histogram(aes(x = bin_median, 
                     y= ..count../sum(..count..)),
                 fill = "white",
                 color = "black") +
  xlab("GC normalized median bin count") +
  ylab("Frequency")
```

### Reads

```{r}
dat_stats %>% 
  gather(key = "parameter",
         value = "value",
         -sample_name) %>% 
  dplyr::filter(parameter %in% c("total_reads", "reads_kept"),
                value > 500) %>% 
  ggplot() +
  geom_histogram(aes(x = value,
                     y = ..count../sum(..count..),
                     fill = parameter),
                 color = "black") +
  facet_wrap(vars(parameter)) +
  scale_fill_manual(values = c("#7D3232", "#32324B")) +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 7)) +
  theme(axis.text.x = element_text(angle = 90),
        strip.background = element_rect(fill= NA),
        legend.position = "none") +
  xlab("Reads") +
  ylab("Frequency")
```


### Duplication 

```{r}
p1<-dat_stats %>% mutate(duplicates=round(duplicates,2)) %>% ggplot()+
  geom_histogram(aes(x =duplicates,y= ..count../sum(..count..)),fill = "white",color = "black")+
  xlab("Duplication ratio") +
  ylab("Frequency")

p2 <- dat_stats %>% mutate(duplicates=round(duplicates,2),total_reads=round(total_reads/1e+6,2)) %>% 
  ggplot(aes(total_reads, duplicates)) +
  geom_point(aes(label = sample_name)) +
  #geom_smooth(method = "lm") +
  #viridis::scale_color_viridis(discrete = TRUE) +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)) +
  scale_y_continuous(breaks = scales::pretty_breaks(n = 10)) +
  xlab("Read count per cell(M)") + 
  ylab("duplication ratio")+xlim(0,2)
p2 <- ggplotly(p2)

subplot(p1,p2,
        nrows = 2,
        titleX = TRUE,
        titleY = TRUE)
```



Row {.tabset .tabset-fade .tabset-pills}
-----------------------------------------------------------------------

### Reads information

```{r}
dt_dat_stats <- dat_stats %>% 
  arrange(desc(reads_kept)) %>% 
  mutate(duplicates = scales::percent(round(duplicates,2)))

DT::datatable(dt_dat_stats, options = list(pageLength = 50))
```

### Removed cells information
```{r}
dt_dat_stats_rm <- dat_stats_rm %>% 
  arrange(desc(reads_kept)) %>% 
  mutate(duplicates = scales::percent(round(duplicates,2)))

DT::datatable(dt_dat_stats_rm, options = list(pageLength = 50))
```


Example Profiles
=======================================================================
```{r making_ratio_plots}
plots<-dir_ls(paste0(here(),"/",params$outdir,"/final_result/ratio_plots_CN"))
```

Row {.tabset .tabset-fade}
-----------------------------------------------------------------------

### 1-2
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[1:2])
```

### 3-4
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[3:4])
```

### 5-6
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[5:6])
```

### 7-8
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[7:8])
```

### 9-10
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[9:10])
```

### 11-12
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[11:12])
```

### 13-14
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[13:14])
```

### 15-16
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[15:16])
```

### 17-18
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[17:18])
```

### 19-20
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[19:20])
```

### 21-22
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[21:22])
```

### 23-24
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[23:24])
```

### 25-26
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[25:26])
```

### 27-28
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[27:28])
```

### 29-30
```{r fig.width=10,fig.height=5}
knitr::include_graphics(plots[29:30])
```


About
=======================================================================

Row
-----------------------------------------------------------------------

### Dashboard
This dashboard was developed by Darlan Conterno Minussi and is a part of the Navin Lab copy number pipeline.

### Clustering
Clustering of the data is being performed with UMAP dimension reduction followed by R phenograph.

UMAP is using the following parameters:

- Seed: 42
- Distance: Manhattan
- min_dist: 0
- n_neighbors: 30 (if n_cells < 30 then k = n_cells - 1)

RPhenograph is using the following parameters:

- k: 30 (if n_cells < 30 then k = n_cells - 1)

### Ploidy Inference
Ploidy inference is using the fourier transform method developed by Alexander Davis. For more information please refer to the following publication.
Ref: Soon

### Phylogenetics
Phylogenetics is using the Neighbor-Joining method and the distance matrix is calculated using manhattan distance.

Row
-----------------------------------------------------------------------

### Packages
The following packages were used in this dashboard:

- flexdashboard
- tidyverse
- here
- fs
- cowplot
- DT
- shiny
- janitor
- ape
- ggtree
- ggsci
- plotly
- amap
- paletteer
- scales
- uwot
- dbscan
- ComplexHeatmap
- flsa
- Rphenograph


